pkg/minishift/docker/image/oci_image_handler.go

```diff
@@ -0,0 +1,366 @@
+you may not use this file except in compliance with the License.
+You may obtain a copy of the License at
+
+    http://www.apache.org/licenses/LICENSE-2.0
+
+Unless required by applicable law or agreed to in writing, software
+distributed under the License is distributed on an "AS IS" BASIS,
+WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+See the License for the specific language governing permissions and
+limitations under the License.
+*/
+
+package image
+
+import (
+	"errors"
+	"fmt"
+	"io"
+	"os"
+	"strconv"
+	"time"
+
+	"bytes"
+	"encoding/json"
+	"github.com/briandowns/spinner"
+	"github.com/containers/image/copy"
+	"github.com/containers/image/oci/layout"
+	"github.com/containers/image/signature"
+	"github.com/containers/image/transports/alltransports"
+	"github.com/containers/image/types"
+	"github.com/docker/machine/libmachine/drivers"
+	"github.com/minishift/minishift/pkg/minikube/sshutil"
+	"github.com/minishift/minishift/pkg/util"
+	"github.com/minishift/minishift/pkg/util/filehelper"
+	"golang.org/x/crypto/ssh"
+	"io/ioutil"
+	"path/filepath"
+	"strings"
+)
+
+const (
+	spinnerSpeed = 200 // In milliseconds
+)
+
+// OciImageHandler is an ImageHandler implementation using OCI format to maintain the local cache.
+type OciImageHandler struct {
+	driver               drivers.Driver
+	dockerClientSettings *dockerClientConfig
+}
+
+type dockerClientConfig struct {
+	DockerHost      string
+	DockerCertPath  string
+	DockerTLSVerify bool
+}
+
+type Index struct {
+	Manifests Manifests `json:"manifests"`
+}
+
+type Manifests []Manifest
+
+type Manifest struct {
+	Annotations Annotations `json:"annotations"`
+}
+
+type Annotations struct {
+	Name string `json:"org.opencontainers.image.ref.name"`
+}
+
+// NewOciImageHandler creates a new ImageHandler which stores cached images in OCI format.
+// It takes a reference to a Driver in order to communicate with the VM and Docker and a map containing the environment settings for the Minishift Docker daemon.
+func NewOciImageHandler(driver drivers.Driver, dockerEnv map[string]string) (*OciImageHandler, error) {
+	settings, err := getDockerSettings(dockerEnv)
+	if err != nil {
+		return nil, err
+	}
+	return &OciImageHandler{driver: driver, dockerClientSettings: settings}, nil
+}
+
+// ImportImages imports cached images from the host into the Docker daemon of the VM.
+func (handler *OciImageHandler) ImportImages(config *ImageCacheConfig) error {
+	out := handler.getOutputWriter(config)
+	fmt.Fprintln(out, "-- Importing cached images ...")
+
+	policyContext, err := handler.getPolicyContext()
+	if err != nil {
+		return fmt.Errorf("Error creating security context: %s", err.Error())
+	}
+
+	availableImages, err := handler.getDockerImages()
+	if err != nil {
+		return err
+	}
+
+	for _, imageName := range config.CachedImages {
+		if _, found := availableImages[imageName]; found {
+			continue
+		}
+
+		fmt.Fprintln(out, fmt.Sprintf("   Importing %s", imageName))
```

Comment:
Please use progress dots, as this can take some time...

Reply:
+1